# Data Preprocessing

This notebook prepares the predictive maintenance dataset for machine learning.

The preprocessing pipeline includes:

- Loading the dataset
- Selecting relevant features
- Separating features and target
- Encoding categorical variables
- Train-test splitting
- Handling class imbalance
- Feature scaling
- Saving the processed datasets

The target variable is `Machine failure`.

In [3]:
import pandas as pd
import numpy as np 

from pathlib import Path

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [4]:
raw = pd.read_csv("../data/raw/ai4i2020.csv")

In [5]:
raw.shape

(10000, 14)

In [6]:
raw.head()

,UDI,Product ID,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
0,1,M14860,M,298.1,308.6,1551,42.8,0,0,0,0,0,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0,0,0,0,0,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0,0,0,0,0,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0,0,0,0,0,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0,0,0,0,0,0


In [7]:
raw.columns.tolist()

['UDI',
 'Product ID',
 'Type',
 'Air temperature [K]',
 'Process temperature [K]',
 'Rotational speed [rpm]',
 'Torque [Nm]',
 'Tool wear [min]',
 'Machine failure',
 'TWF',
 'HDF',
 'PWF',
 'OSF',
 'RNF']

In [12]:
features = [
    "Type",
    "Air temperature [K]",
    "Process temperature [K]",
    "Rotational speed [rpm]",
    "Torque [Nm]",
    "Tool wear [min]"
]

target = "Machine failure"

In [13]:
df = raw[features + [target]].copy()

df.head()

,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure
0,M,298.1,308.6,1551,42.8,0,0
1,L,298.2,308.7,1408,46.3,3,0
2,L,298.1,308.5,1498,49.4,5,0
3,L,298.2,308.6,1433,39.5,7,0
4,L,298.2,308.7,1408,40.0,9,0


In [15]:
df.shape
df.info()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Type                     10000 non-null  object 
 1   Air temperature [K]      10000 non-null  float64
 2   Process temperature [K]  10000 non-null  float64
 3   Rotational speed [rpm]   10000 non-null  int64  
 4   Torque [Nm]              10000 non-null  float64
 5   Tool wear [min]          10000 non-null  int64  
 6   Machine failure          10000 non-null  int64  
dtypes: float64(3), int64(3), object(1)
memory usage: 547.0+ KB


Type                       0
Air temperature [K]        0
Process temperature [K]    0
Rotational speed [rpm]     0
Torque [Nm]                0
Tool wear [min]            0
Machine failure            0
dtype: int64

In [16]:
X = df.drop(columns=["Machine failure"])
y = df["Machine failure"]

In [17]:
X.head()

,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min]
0,M,298.1,308.6,1551,42.8,0
1,L,298.2,308.7,1408,46.3,3
2,L,298.1,308.5,1498,49.4,5
3,L,298.2,308.6,1433,39.5,7
4,L,298.2,308.7,1408,40.0,9


In [18]:
y.value_counts()

Machine failure
0    9661
1     339
Name: count, dtype: int64

In [19]:
X = pd.get_dummies(
    X,
    columns=["Type"],
    drop_first=True,
    dtype=int
)

In [20]:
X.head()

,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Type_L,Type_M
0,298.1,308.6,1551,42.8,0,0,1
1,298.2,308.7,1408,46.3,3,1,0
2,298.1,308.5,1498,49.4,5,1,0
3,298.2,308.6,1433,39.5,7,1,0
4,298.2,308.7,1408,40.0,9,1,0


In [21]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 10000 entries, 0 to 9999
Series name: Machine failure
Non-Null Count  Dtype
--------------  -----
10000 non-null  int64
dtypes: int64(1)
memory usage: 78.3 KB


In [22]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

In [23]:
print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)


X_train: (8000, 7)
X_test: (2000, 7)
y_train: (8000,)
y_test: (2000,)


In [24]:
print("Training set:")
print(y_train.value_counts())
print(y_train.value_counts(normalize=True))

print("\nTest set:")
print(y_test.value_counts())
print(y_test.value_counts(normalize=True))

Training set:
Machine failure
0    7729
1     271
Name: count, dtype: int64
Machine failure
0    0.966125
1    0.033875
Name: proportion, dtype: float64

Test set:
Machine failure
0    1932
1      68
Name: count, dtype: int64
Machine failure
0    0.966
1    0.034
Name: proportion, dtype: float64


In [25]:
from sklearn.preprocessing import StandardScaler

numeric_features = [
    "Air temperature [K]",
    "Process temperature [K]",
    "Rotational speed [rpm]",
    "Torque [Nm]",
    "Tool wear [min]"
]

scaler = StandardScaler()

X_train[numeric_features] = scaler.fit_transform(
    X_train[numeric_features]
)

X_test[numeric_features] = scaler.transform(
    X_test[numeric_features]
)

In [26]:
X_train[numeric_features].describe()

,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min]
count,8.000000e+03,8.000000e+03,8.000000e+03,8.000000e+03,8.000000e+03
mean,1.758593e-15,1.053468e-14,3.010925e-16,-1.474376e-16,-3.064216e-17
std,1.000063e+00,1.000063e+00,1.000063e+00,1.000063e+00,1.000063e+00
min,-2.356591e+00,-2.910801e+00,-2.052017e+00,-3.613500e+00,-1.692947e+00
25%,-8.541261e-01,-8.152711e-01,-6.484822e-01,-6.790515e-01,-8.597185e-01
50%,4.735268e-02,6.349964e-02,-2.008982e-01,9.645501e-03,-1.076908e-02
75%,7.485028e-01,7.394771e-01,4.069319e-01,6.783803e-01,8.696230e-01
max,2.250967e+00,2.564616e+00,7.441184e+00,3.652753e+00,2.284539e+00


In [27]:
X_test[numeric_features].describe()

,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min]
count,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000
mean,-0.001302,-0.001698,-0.016046,-0.008211,0.020909
std,1.008835,1.014772,0.952336,0.974761,1.003447
min,-2.356591,-2.843204,-1.864142,-3.413877,-1.692947
25%,-0.804044,-0.747673,-0.631905,-0.679052,-0.843997
50%,0.047353,0.063500,-0.200898,-0.000336,0.067837
75%,0.748503,0.739477,0.384829,0.648437,0.838180
max,2.200885,2.429421,7.104114,3.193622,2.174490


In [28]:
from pathlib import Path

processed_path = Path("../data/processed")
processed_path.mkdir(parents=True, exist_ok=True)

In [29]:
X_train.to_csv(processed_path / "X_train.csv", index=False)
X_test.to_csv(processed_path / "X_test.csv", index=False)

y_train.to_csv(processed_path / "y_train.csv", index=False)
y_test.to_csv(processed_path / "y_test.csv", index=False)

In [30]:
print("Processed files saved successfully.")

print(list(processed_path.iterdir()))

Processed files saved successfully.
[PosixPath('../data/processed/X_train.csv'), PosixPath('../data/processed/y_train.csv'), PosixPath('../data/processed/y_test.csv'), PosixPath('../data/processed/X_test.csv')]


In [31]:
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

print("\nMissing values:")
print("X_train:", X_train.isnull().sum().sum())
print("X_test:", X_test.isnull().sum().sum())
print("y_train:", y_train.isnull().sum())
print("y_test:", y_test.isnull().sum())

print("\nTarget distribution:")
print("Training:")
print(y_train.value_counts())

print("\nTesting:")
print(y_test.value_counts())

print("\nFeature columns:")
print(X_train.columns.tolist())

X_train shape: (8000, 7)
X_test shape: (2000, 7)
y_train shape: (8000,)
y_test shape: (2000,)

Missing values:
X_train: 0
X_test: 0
y_train: 0
y_test: 0

Target distribution:
Training:
Machine failure
0    7729
1     271
Name: count, dtype: int64

Testing:
Machine failure
0    1932
1      68
Name: count, dtype: int64

Feature columns:
['Air temperature [K]', 'Process temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]', 'Tool wear [min]', 'Type_L', 'Type_M']


In [32]:
import joblib
from pathlib import Path

models_path = Path("../models")
models_path.mkdir(parents=True, exist_ok=True)

joblib.dump(
    scaler,
    models_path / "scaler.joblib"
)

print("Scaler saved successfully.")

Scaler saved successfully.


# Day 2 Complete — Data Preprocessing

The equipment sensor dataset was successfully prepared for machine learning.

### Completed Tasks
- Selected relevant sensor and machine features.
- Excluded identifier and failure-mode columns to avoid target leakage.
- Separated input features (`X`) from the target (`y`).
- Converted the categorical `Type` feature using one-hot encoding.
- Performed an 80/20 stratified train-test split.
- Preserved the original failure-class distribution in both datasets.
- Standardized numerical features using `StandardScaler`.
- Fitted the scaler only on the training data to prevent data leakage.
- Saved the processed training and testing datasets to `data/processed/`.

### Final Dataset
- Training samples: 8,000
- Testing samples: 2,000
- Input features: 7
- Target: `Machine failure`
- Training failures: 271
- Testing failures: 68

The preprocessed data is now ready for machine-learning model development.